Add gnosis transaction preview dialog

chinins opened the pull request:
## Description

This PR adds transaction preview logic for the gnosis safe control dialog. I've already added preview for transfer funds & for raw transaction. NFT & arbitrary transactions are not finished so there is a separate issue to add them when they are.

I've also added a fix for another issue - validate on load.

I've also added `TransactionTypes` enum across the dialog.
<img width="551" alt="Screenshot 2022-07-22 at 13 37 27" src="https://user-images.githubusercontent.com/34057551/180422038-a09a57bb-0fff-40b2-8645-3018d354995f.png">
<img width="546" alt="Screenshot 2022-07-22 at 13 37 56" src="https://user-images.githubusercontent.com/34057551/180422139-1648b726-8ea9-44f3-bcf6-a50449a17fe6.png">



Resolves #3543 
Resolves #3664

ArmandoGraterol reviewed (changes requested):
Great work! Just a couple of details/questions and this should be good to go from my side.

1. In the specs now the "Target contract" field is called "Safe".
![FireShot Capture 674 - Safe Control – Figma - www figma com](https://user-images.githubusercontent.com/18473896/180580112-0ec77df0-8f20-4b99-9d5d-5394501a83a3.png)

2. Also it seems like the values of the things like `Data` or `Value (wei)` should be bolder. In the specs they are `600`, which looks too different from the `400` we have implemented.

3. "Title of the set of all the arbitrary transactions" should now be simply "Title"

4. "Explain why you’re making arbitrary transaction (optional)" is now "Description" which I'm gonna assume now that it's required because it doesn't have the "(optional)"

5. The field values should be aligned to the right, as per the screenshot above of the specs.

6. If I focus on the title field, then focus out and go back to the previous screen then the "Create transaction" button shows as disabled even though the current form is completed.

7. In the case of the "Transfer funds" transaction form, when I go back the amount field is empty and I have to re-enter the amount.

8. This is more of an idea/suggestion, what do you think of adding `appareance={{ theme: 'sidepadding' }}` to the section with the transaction title? In my opinion it seems like it makes it closer to the specs (in the specs there's not that much space between the title and description).

willm30 reviewed (commented):
Correct me if I'm wrong but this now needs to be updated to reflect the added ability to add multiple transactions? I just added 3 but only received a preview for the first one.

willm30 commented:
> Correct me if I'm wrong but this now needs to be updated to reflect the added ability to add multiple transactions? I just added 3 but only received a preview for the first one.

Ah, I see this is a [separate issue ](https://github.com/JoinColony/colonyDapp/issues/3564)👍

Julianahlk commented on src/modules/dashboard/components/Dialogs/GnosisControlSafeDialog/SafeTransactionPreview.tsx:
```diff
@@ -0,0 +1,205 @@
+  },
+  explainWhy: {
+    id: 'dashboard.GnosisControlSafeDialog.SafeTransactionPreview.explainWhy',
+    defaultMessage: `Explain why you’re making arbitrary transaction (optional)`,
```
I think it would sound better if we add the word "the": "Explain why you’re making the arbitrary transaction (optional)"

Julianahlk replied:
Never mind, I was going through your code and while I was reviewing it you already made the changes.

chinins commented:
> Great work! Just a couple of details/questions and this should be good to go from my side.
> 
> 1. In the specs now the "Target contract" field is called "Safe".

> 2. Also it seems like the values of the things like `Data` or `Value (wei)` should be bolder. In the specs they are `600`, which looks too different from the `400` we have implemented.
> 3. "Title of the set of all the arbitrary transactions" should now be simply "Title"
> 4. "Explain why you’re making arbitrary transaction (optional)" is now "Description" which I'm gonna assume now that it's required because it doesn't have the "(optional)"
> 5. The field values should be aligned to the right, as per the screenshot above of the specs.
> 6. If I focus on the title field, then focus out and go back to the previous screen then the "Create transaction" button shows as disabled even though the current form is completed.
> 7. In the case of the "Transfer funds" transaction form, when I go back the amount field is empty and I have to re-enter the amount.
> 8. This is more of an idea/suggestion, what do you think of adding `appareance={{ theme: 'sidepadding' }}` to the section with the transaction title? In my opinion it seems like it makes it closer to the specs (in the specs there's not that much space between the title and description).

@ArmandoGraterol, thank you for the feedback Please, see my comments:
1, 3 & 4 - updated (didn't notice the spec changed before). Descriptions should still be optional (saying it in brackets) - confirmed with Arren
2 - in the figma it says font-weight: 400 but I asked Arren & confirmed that we can do 700 - updated.
5 - on my screen it's all aligned to the right like in the specs. Can you please check and maybe offer a fix for your system (can't fix myself as I can't replicate)
<img width="499" alt="Screenshot 2022-07-26 at 11 59 49" src="https://user-images.githubusercontent.com/34057551/181032604-d7081483-213f-49dd-aa53-db207ecc1e49.png">

7 -  good catch. Fixed now although it take me a while 🙈
8 - talked to Arren and he said use margin above the Title field but not between  title and description - updated.

6 - OMG, this is a nightmare bug! I understand why it's happening but I haven't been able to fix it. Kept trying for almost the whole day. It validates it based on the old validation form although it's already update and it triggers re-validation only after you focus & unfocus a field. I tried all of the Formik validation props & functions but unsuccessfully. I will try again tomorrow... Any suggestions are welcome.

Julianahlk commented:
I noticed that the avatar does not have a border radius.
<img width="497" alt="Screen Shot 2022-07-26 at 5 25 23 PM" src="https://user-images.githubusercontent.com/71563622/181037862-71ea9f08-fcfd-454a-bc21-0e4bd51e211b.png">

Julianahlk commented on src/modules/dashboard/components/Dialogs/GnosisControlSafeDialog/SafeTransactionPreview.tsx:
```diff
@@ -0,0 +1,205 @@
+  },
+  transactionsSetTitle: {
+    id: `dashboard.GnosisControlSafeDialog.SafeTransactionPreview.transactionsSetTitle`,
+    defaultMessage: 'Title of the set of all the arbitrary transactions',
```
Never mind, I was going through your code and while I was reviewing it you already made the changes.

ArmandoGraterol commented:
> 1, 3 & 4 - updated (didn't notice the spec changed before). Descriptions should still be optional (saying it in brackets) - confirmed with Arren

Thank you for confirming that!

> 5 - on my screen it's all aligned to the right like in the specs. Can you please check and maybe offer a fix for your system (can't fix myself as I can't replicate) <img alt="Screenshot 2022-07-26 at 11 59 49" width="499" src="https://user-images.githubusercontent.com/34057551/181032604-d7081483-213f-49dd-aa53-db207ecc1e49.png">

My bad, I think I didn't explain myself well, I'm talking about the inner alignment of the values:
![FireShot Capture 678 - Actions - Colony - wonker - localhost](https://user-images.githubusercontent.com/18473896/181131057-3c5c1de7-4d74-4532-bd3e-63d34c77ace6.png)

Data's value is internally aligned to the left when it has multiple lines as you can see from my screenshot, while in the specs even internally they are aligned to the right:

![180580112-0ec77df0-8f20-4b99-9d5d-5394501a83a3](https://user-images.githubusercontent.com/18473896/181131198-eb79d423-cb55-48f3-b8fc-dc9659440623.png)


> 6 - OMG, this is a nightmare bug! I understand why it's happening but I haven't been able to fix it. Kept trying for almost the whole day. It validates it based on the old validation form although it's already update and it triggers re-validation only after you focus & unfocus a field. I tried all of the Formik validation props & functions but unsuccessfully. I will try again tomorrow... Any suggestions are welcome.

I'll also take a stab at it to see if I can come up with some suggestions :+1:

ArmandoGraterol commented:
@chinins Suggestion time! This worked for me:

1. Make the initial value of `showPreview` `null`: `const [showPreview, setShowPreview] = useState<boolean | null>(null);`.
2. Add this effect to the `GnosisControlSafeForm` component:
```
  useEffect(() => {
    if (!isNil(showPreview) && !showPreview) {
      validateForm();
    } 
  }, [showPreview, validateForm]);
```

`validateForm` comes from the `FormikProps`.

This way I have been able to re-run the validation when going back to the previous dialog and get rid of the preview error :) Perhaps there's a better way but at least it's a start.

chinins commented:
> Data's value is internally aligned to the left when it has multiple lines as you can see from my screenshot, while in the specs even internally they are aligned to the right:

I see now - when the data has spaces! Good catch - fixed now.

chinins commented:
> @chinins Suggestion time! This worked for me:
> 
> 1. Make the initial value of `showPreview` `null`: `const [showPreview, setShowPreview] = useState<boolean | null>(null);`.
> 2. Add this effect to the `GnosisControlSafeForm` component:
> 
> ```
>   useEffect(() => {
>     if (!isNil(showPreview) && !showPreview) {
>       validateForm();
>     } 
>   }, [showPreview, validateForm]);
> ```
> 
> `validateForm` comes from the `FormikProps`.
> 
> This way I have been able to re-run the validation when going back to the previous dialog and get rid of the preview error :) Perhaps there's a better way but at least it's a start.

I think it's a very good solution. Yesterday everything I tried was in the Dialog component (even the validate).
I've added the fix. It seems that the null & isNil check is not even necessary. Working without it.

Thank you for the idea!

willm30 reviewed (commented):
Looking good. One question: should there be validation on the title field? If you enter a single space it will let you submit the form without a title.

ArmandoGraterol reviewed (approved):
Looks good to me! Great job. Just take into consideration what Will mentioned about the title

chinins commented:
@willm30 @ArmandoGraterol it's the same validation as we have everywhere else in the dapp - empty strings are accepted.
